### PR TITLE
Large refactor of the core structs and encryption model

### PR DESCRIPTION
**Pull request by mhammond**
As discussed a little in person - this is a large patch that changes the shape of the structs etc, but does *not* radically change the shape of the API (other than to require the encryption key where necessary) - those API changes should be done ASAP, but probably in a different PR.

I'm open to all bike-shedding - some of the struct names don't seem ideal, but better alternative aren't completely obvious. YMMV, so if you think better names are obvious, then please share!

Except for migration, all tests work, and all `TODO-sqlcipher` comments have been removed. Android compiles. I didn't merge Sammy's migration PR yet because this PR adds some primitives that could be used (and if not for needing to migrate the mirror, that migration could almost entirely be implemented via `import_multiple`.

### CODE REVIEW

File: components/logins/src/logins.udl
Comment (mhammond) [inline components/logins/src/logins.udl:15]: actually, this can probably die as `Login` has a method that does the same thing.

File: components/logins/src/db.rs
Comment (mhammond) [inline components/logins/src/db.rs:337]: The "OR IGNORE" previously here seems like a footgun, so I removed it.

File: components/logins/src/db.rs
Comment (mhammond) [inline components/logins/src/db.rs:375]: ditto here - no "OR" clauses because all callers of this should be 100% confident there's no existing record with the same guid.

File: components/logins/src/db.rs
Comment (mhammond) [inline components/logins/src/db.rs:435]: stuff got a but messy because this public function wants to add new records via a `Login` struct rather than an `UpdatableLogin` (eg, `validate_and_fixup` for a `Login` struct is only needed because this exists). This was the thing that pushed me over the edge of creating the `LoginFields` struct rather than flattening them.

File: components/logins/src/login.rs
Comment (bendk) [inline components/logins/src/login.rs:536]: I'm not sure if this should be implemented on the `Login` struct since there's no way to validate the encrypted fields.  It seems better to validate the `LoginFields`, then decrypt `enc_fields` and use the `EncryptedLogin.validate_and_fixup()` code.

**Review by bendk — COMMENTED** [review]
I like the overall approach here.  The only question I had was the validation code

I've gone forth in my mind on nested vs non-nested structs and at this point I think a) there are a bunch of minor differences that mostly balance each other out and b) I think I slightly prefer the nested approach.

File: components/logins/src/db.rs
Comment (lougeniaC64) [reply components/logins/src/db.rs:551]: I agree about removing the dupe. The only caveat I'd have is considering what would happen if the dupe has an older `time_password_changed` timestamp. In that case we probably want to keep the dupe or at minimum the dupe's metadata.

File: examples/sync-pass/src/sync-pass.rs
Comment (lougeniaC64) [inline examples/sync-pass/src/sync-pass.rs:92]: nit: Seeing `EncryptedFields` outside of the `Login` struct makes me want it to be renamed `EncryptedLoginFields` so this struct isn't construed as a generic struct not associated with logins. Although perhaps it would read as redundant nested inside of the `Login` struct but maybe worth it? Do what you will with this.

File: components/logins/src/db.rs
Comment (lougeniaC64) [inline components/logins/src/db.rs:255]: nit: `login.fields.origin` instead of `fields.origin`

File: components/logins/src/logins.udl
Comment (mhammond) [reply components/logins/src/logins.udl:15]: well, it does in rust, but it's a dictionary in UDL, so this needs to remain.

File: components/logins/src/db.rs
Comment (mhammond) [reply components/logins/src/db.rs:435]: I'm not sure what you are asking here, but to restate what you said, validating a `Login` does just validate the `fields` and `enc_fields` members. This means there's no validation on the metadata, which possibly should happen, but the existing code didn't check that either IIUC.

File: components/logins/src/db.rs
Comment (mhammond) [reply components/logins/src/db.rs:551]: I think an important consideration here will be the existing Fenix code - eg, will the UI update correctly if we do that deletion? It would clearly be bad if the login we removed stayed in the list of logins. Sadly I can't build Fenix currently for reasons I don't understand, so can't check.

The metadata is an interesting question - maybe we should delete the one being updated as that has entirely wrong metadata now that one of the keys has changed.

File: components/logins/src/login.rs
Comment (mhammond) [reply components/logins/src/login.rs:541]: I like `SecureFields` - I'll make that change tomorrow!

Comment (mhammond) [reply components/logins/src/login.rs:541]: Or as Lougenia comments below, `SecureLoginFields`? Or `LoginSecureFields` (doesn't read as well, but is more consistent!)? WDYT?

File: components/logins/src/login.rs
Comment (mhammond) [reply components/logins/src/login.rs:550]: hrm, interesting.

File: components/logins/src/db.rs
Comment (mhammond) [reply components/logins/src/db.rs:609]: yes, nice catch, thanks

File: components/logins/src/login.rs
Comment (mhammond) [reply components/logins/src/login.rs:536]: hrm, yeah, that's probably correct.

File: components/logins/src/login.rs
Comment (bendk) [reply components/logins/src/login.rs:541]: I like the readability of `SecureLoginFields`

File: components/logins/src/login.rs
Comment (mhammond) [reply components/logins/src/login.rs:550]: I really low-cost thing was to just `[#serde(rename="...")]` the fields, so it is `{"u": "username", "p":"password"}` - not as tight as a custom serializer could do, but probably good enough - WDYT?

File: components/logins/src/login.rs
Comment (mhammond) [inline components/logins/src/login.rs:545]: Note to self: actually check this!

**Comment by mhammond** [comment]
I pushed up the rename to `SecureLoginFields` and a few other changes. I've a few other comments to address before I'll re-request review, but I think this is in good enough shape to unblock the Fenix and iOS efforts?

File: components/logins/src/login.rs
Comment (bendk) [reply components/logins/src/login.rs:550]: Yeah, I think that's a good balance between simplicity and reducing the field size.

**Comment by mhammond** [comment]
After doing all review comments I merged main into logins-per-field-encryption, then rebased this branch against that. I think this is ready to roll now, so re-flagging reviews.

**Review by bendk — APPROVED** [review]
Looks good to me, we can keep iterating in the main branch.

**Comment by bendk** [comment]
I think we should consider renaming the `UpdatableLogin` struct.  We use it for both `add()` and `update()` which is a bit confusing.  What able `SavableLogin` or `MutableLogin`?

**Review by lougeniaC64 — APPROVED** [review]
LGTM!

**Comment by mhammond** [comment]
> I think we should consider renaming the `UpdatableLogin` struct. We use it for both `add()` and `update()` which is a bit confusing. What able `SavableLogin` or `MutableLogin`?

Not a huge fan of either of those names TBH and very lightly prefer `UpdatableLogin` over them. How about `Editable`?

**Comment by mhammond** [comment]
> How about `Editable`?

`Writable`?

(Don't really care, but worth trying to get something we all like)

**Comment by bendk** [comment]
`Editable` or `Writable` are both good with me.

**Comment by bendk** [comment]
I'm for merging this one into `logins-per-field-encryption` ASAP and maybe also merging the extra commits here: https://github.com/bendk/application-services/tree/logins-per-field-encryption.  I feel like merging frequently is the best way to deal with the API being in flux.  Definitely don't let my comments abut the struct names slow this down.

**Pull request merged**
